File: leetcode/0152-maximum-product-subarray/solution.py
from typing import List


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        result = nums[0]
        curr_max = curr_min = nums[0]

        for num in nums[1:]:
            candidates = (num, num * curr_max, num * curr_min)
            curr_max = max(candidates)
            curr_min = min(candidates)
            result = max(result, curr_max)

        return result

File: leetcode/0152-maximum-product-subarray/Solution.java
class Solution {
    public int maxProduct(int[] nums) {
        int result = nums[0];
        int currMax = nums[0], currMin = nums[0];

        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];
            int candidate1 = num * currMax;
            int candidate2 = num * currMin;

            currMax = Math.max(num, Math.max(candidate1, candidate2));
            currMin = Math.min(num, Math.min(candidate1, candidate2));

            result = Math.max(result, currMax);
        }

        return result;
    }
}
